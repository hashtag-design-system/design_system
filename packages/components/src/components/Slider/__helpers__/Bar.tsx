import React from "react";
import { SliderFProps } from "../index";

export const Bar = React.forwardRef<HTMLSpanElement, Pick<SliderFProps, "className">>(({ className }, ref) => {
  return (
    <span className={`slider__bar shadow__inset-sm ${className}`}>
      <span ref={ref} className={`slider__bar__progress ${className}`} />
    </span>
  );
});
