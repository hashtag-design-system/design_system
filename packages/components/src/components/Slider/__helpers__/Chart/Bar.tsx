import React from "react";

type Props = Pick<React.CSSProperties, "height">;

export const Bar: React.FunctionComponent<Props> = ({ height = 50 }) => {
  return (
    <div className="slider__chart__bar" style={{ height: `${height}%` }} data-testid="slider-chart-bar"></div>
  );
};
