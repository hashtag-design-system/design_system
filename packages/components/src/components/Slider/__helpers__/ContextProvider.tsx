import React, { useCallback } from "react";
import { SliderProps } from "..";
import { SliderContextProvider, SliderContextType } from "../../../utils/contexts/SliderContext";

export type Props = {
  value: number;
  setValue: React.Dispatch<React.SetStateAction<number>>;
} & Omit<SliderProps, "marks" | "lockOnMarks" | "zeroPercentageOnEdgeMarks">;

export const ContextProvider: React.FC<Props> = ({
  value,
  setValue,
  min = 0,
  max = 100,
  defaultValue = max / 2,
  step = 1,
  children,
}) => {
  // const incr = () => {
  //   return value + step;
  // };

  // const dcr = () => {
  //   return value - step;
  // };

  // Calculate the percentage of the value
  const calcPercentage = useCallback(
    (markValue?: number): number => {
      const num = markValue !== undefined ? markValue : value;
      return ((num - min) / (max - min)) * 100;
    },
    [value, min, max]
  );

  // Calculate the value from the percentage
  const calcValue = useCallback(
    (percentage: number): number => {
      return (percentage / 100) * max;
    },
    [max]
  );

  // const calcPosition = (markValue: number, idx: number, last: boolean): number | undefined => {
  //   let value: number = calcValue(markValue);
  //   if (!last) {
  //     if (zeroPercentageOnEdgeMarks && idx === 0) {
  //       value = 0;
  //     } else {
  //       value = value - (window.outerWidth >= 700 ? 1 : 1.75);
  //     }
  //     return value;
  //   } else {
  //     return undefined;
  //   }
  // };

  const providerValue: SliderContextType = {
    min,
    max,
    defaultValue,
    step,
    value,
    setValue,
    calcValue,
    calcPercentage,
  };

  return <SliderContextProvider value={providerValue}>{children}</SliderContextProvider>;
};
