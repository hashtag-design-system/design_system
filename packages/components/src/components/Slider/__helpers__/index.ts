export * from "./Bar/Bar";
export * from "./Chart/Chart";
export * from "./Marks/Marks";
export * from "./Thumb/Thumb";
