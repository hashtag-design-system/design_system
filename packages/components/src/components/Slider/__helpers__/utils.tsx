import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { SliderContextProvider } from "../../../utils/contexts/SliderContext";
import { SliderFProps } from "../index";

export const defaultProps: SliderFProps = {
  min: 0,
  max: 100,
  step: 1,
  marks: [
    { value: 0, label: "0" },
    { value: 100, label: "100" },
  ],
  lockOnMarks: false,
  zeroPercentageOnEdgeMarks: false,
};

type SliderCustomRenderOptions = {
  providerProps: SliderFProps;
  renderOptions?: RenderOptions;
};

export const sliderCustomRender = (ui: React.ReactElement, options?: SliderCustomRenderOptions) => {
  return render(
    <SliderContextProvider value={{ ...defaultProps, ...options?.providerProps, calcPercentage: () => 1, calcValue: () => 1 }}>
      {ui}
    </SliderContextProvider>,
    options?.renderOptions
  );
};
