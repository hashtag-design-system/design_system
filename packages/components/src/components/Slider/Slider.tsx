import React, { useCallback, useEffect, useRef, useState } from "react";
import { calculatePercentage, calculateValue } from "../../utils";
import { useClassnames, useWindowDimensions } from "../../utils/hooks";
import Input, { BaseReactInputHTMLAttributes, NumberInputProps } from "../Input";
import Double from "./Double";

// See -> https://www.youtube.com/watch?v=zOA2vpx44Nw
// See -> https://www.youtube.com/watch?v=mvq8uOGFqlc
// See -> https://www.youtube.com/watch?v=MxbEjINYIPc

export type SliderThumbProp = { defaultValue?: number };
export type SliderMarkProp = { value: number; label?: string };
type SliderChartDataProp = { value: number };

export type Props = Omit<NumberInputProps, "value" | "defaultValue"> & {
  thumb?: SliderThumbProp;
  marks: SliderMarkProp[];
  lockOnMarks?: boolean;
  zeroPercentageOnEdgeMarks?: boolean;
  formatRegExp?: {
    searchValue: string | RegExp;
    replaceValue: string;
  };
  chart?: {
    type: "bar";
    data: SliderChartDataProp[];
    percentage?: boolean;
  };
};

interface SubComponents {
  Double: typeof Double;
}
const DEFAULT_SIZE = 0.875;

const Slider: React.FC<
  Omit<React.ComponentPropsWithRef<"input">, "value" | "defaultValue"> &
    Props &
    Omit<BaseReactInputHTMLAttributes, "value" | "defaultValue">
> &
  SubComponents = ({
  min = 0,
  max = 100,
  thumb = { defaultValue: max / 2 },
  step = 1,
  marks,
  lockOnMarks = false,
  zeroPercentageOnEdgeMarks = false,
  inchange,
  formatRegExp,
  chart,
  ref,
  ...props
}) => {
  const [classNames, rest] = useClassnames("slider shadow__inset-sm", props);
  const [value, setValue] = useState<number>(thumb.defaultValue!);
  const [prevKey, setPrevKey] = useState<string>("0");
  const [onHover, setOnHover] = useState<boolean>(false);
  const [size, setSize] = useState<number>(DEFAULT_SIZE);
  const [sliderStep, setSliderStep] = useState<number>(step);
  const [chartOverlayLeft, setChartOverlayLeft] = useState<number>(0);
  const windowDimensions = useWindowDimensions();

  const progressRef = useRef<HTMLSpanElement>(null);
  const chartFrameRef = useRef<HTMLDivElement>(null);
  const chartOverlayFrameRef = useRef<HTMLDivElement>(null);
  const chartOverlay = useRef<HTMLDivElement>(null);

  const incr = () => {
    setValue(value => value + sliderStep);
  };

  const dcr = () => {
    setValue(value => value - sliderStep);
  };

  const calcPercentage = useCallback(
    (number?: number): number => {
      const num = number === undefined ? value : number;
      const res = calculatePercentage(num, min, max);
      return res;
    },
    [value, min, max]
  );

  const calcValue = useCallback(
    (percentage: number): number => {
      const res = calculateValue(percentage, max);
      return res;
    },
    [max]
  );

  const calcPosition = (mark: SliderMarkProp, last: boolean): number | undefined => {
    const { value: markValue, label } = mark;
    if (!last) {
      return calcPercentage(markValue) - 0.55 * String(label !== undefined ? label : markValue).length;
    } else if (markValue === min) {
      return 0;
    } else {
      return undefined;
    }
  };

  const calcBarHeight = useCallback(
    (data: SliderChartDataProp) => {
      if (!chart) {
        return undefined;
      }
      const { percentage } = chart;
      const { value } = data;
      if (percentage) {
        return value;
      } else {
        const data = chart.data.map(datum => datum.value);
        const maxVal = Math.max(...data);
        const perc = calculatePercentage(value, 0, maxVal, { returnRounded: true });
        return perc;
      }
    },
    [chart]
  );

  const setStep = (valueAsNumber: number) => {
    const currentValueIdx = marks.map(mark => mark.value).indexOf(value);
    const nextValue = marks[currentValueIdx + 1];
    const prevValue = marks[currentValueIdx - 1];
    let targetValue = value;
    if (nextValue === undefined) {
      targetValue = max;
    } else if (prevValue === undefined) {
      targetValue = 0;
    }

    if (valueAsNumber > targetValue) {
      const val = nextValue.value;
      setSliderStep(val - targetValue);
    } else {
      const val = prevValue.value;
      setSliderStep(targetValue - val);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    e.preventDefault();
    const { key } = e;

    const currentValueIdx = marks.map(mark => mark.value).indexOf(value);
    if (key === "ArrowRight" || key === "ArrowUp") {
      if (lockOnMarks && currentValueIdx !== -1) {
        const nextValue = marks[currentValueIdx + 1];

        if (nextValue) {
          const val = nextValue.value;
          setSliderStep(val - value);
          setValue(val);
        }
      } else {
        incr();
      }
    } else if (key === "ArrowLeft" || key === "ArrowDown") {
      if (lockOnMarks && currentValueIdx !== -1) {
        const prevValue = marks[currentValueIdx - 1];

        if (prevValue) {
          const val = prevValue.value;
          setSliderStep(value - val);
          setValue(val);
        }
      } else {
        dcr();
      }
    }

    // Move thumb and change value, directly from numeric keyboard
    if (!isNaN(+key)) {
      const numKey = parseInt(key);
      const number = calcValue(numKey * 10);
      if (numKey === 0 && prevKey === "1") {
        setValue(max);
      } else {
        setValue(number);
      }
    }
    setPrevKey(key);
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { valueAsNumber } = e.target;

    if (lockOnMarks) {
      const closest = marks
        .map(mark => mark.value)
        .reduce(function (prev, curr) {
          return Math.abs(curr - valueAsNumber) < Math.abs(prev - valueAsNumber) ? curr : prev;
        });

      const marksValues = marks.map(mark => mark.value);

      setStep(valueAsNumber);

      if (marksValues.includes(valueAsNumber)) {
        setValue(valueAsNumber);
      } else {
        setValue(closest);
      }
    } else {
      setValue(valueAsNumber);
    }

    if (inchange) {
      inchange(valueAsNumber);
    }
  };

  useEffect(() => {
    if (progressRef && progressRef.current) {
      progressRef.current.style.width = `${calcPercentage()}%`;
    }
  }, [calcPercentage]);

  useEffect(() => {
    if (onHover) {
      const valueLength = String(max).length;
      let newValue: number = DEFAULT_SIZE * 1.75;

      if (valueLength > 3) {
        newValue = newValue + (valueLength - 3) * 0.4;
      }

      setSize(newValue);
    } else {
      setSize(DEFAULT_SIZE);
    }
  }, [max, onHover, size]);

  useEffect(() => {
    if (chartFrameRef && chartFrameRef.current && chartOverlayFrameRef && chartOverlayFrameRef.current) {
      chartOverlayFrameRef.current.style.width = `${chartFrameRef.current.offsetWidth}px`;
    }
  }, [windowDimensions.width, chartFrameRef, chartOverlayFrameRef]);

  useEffect(() => {
    if (chartOverlay && chartOverlay.current) {
      setChartOverlayLeft(chartOverlay.current.offsetLeft);
    }
  }, [chartOverlay]);

  return (
    <div className="slider__wrapper flex-column-flex-start-center">
      {chart && (
        <div className="slider__chart flex-column-flex-start-flex-start" data-onhover={onHover}>
          <div className="slider__chart__overlay" style={{ right: `${calcPercentage(max - value)}%` }} ref={chartOverlay}>
            <div
              className="slider__chart__frame slider__chart__frame__overlay"
              ref={chartOverlayFrameRef}
              style={{ left: `-${chartOverlayLeft}px` }}
            >
              {chart.data.map((datum, i) => {
                return <div key={i} className="slider__chart__bar" style={{ height: `calc(${calcBarHeight(datum)}%)` }}></div>;
              })}
            </div>
          </div>
          <div className="slider__chart__frame" ref={chartFrameRef}>
            {chart.data.map((datum, i) => {
              return <div key={i} className="slider__chart__bar" style={{ height: `calc(${calcBarHeight(datum)}%)` }}></div>;
            })}
          </div>
        </div>
      )}
      <div className="slider__field flex-column-flex-start-stretch" onKeyDown={e => handleKeyDown(e)}>
        <Input.BaseInput
          type="range"
          min={min}
          max={max}
          step={sliderStep}
          value={value}
          className={classNames}
          onChange={e => handleChange(e)}
          ref={ref}
          tabIndex={-1}
          role="slider"
          aria-valuemin={min}
          aria-valuemax={max}
          aria-valuenow={value}
          onMouseOver={() => setOnHover(true)}
          onMouseOut={() => setOnHover(false)}
          onTouchMove={() => setOnHover(true)}
          onTouchEnd={() => setOnHover(false)}
          {...rest}
        >
          <span className="slider__bar shadow__inset-sm">
            <span ref={progressRef} className="slider__bar__progress" />
          </span>
        </Input.BaseInput>
        <div
          className="slider__thumb"
          style={{
            left: `${calcPercentage()}%`,
            width: `${size}em`,
            height: `${size}em`,
          }}
          tabIndex={0}
          data-onhover={onHover}
        >
          <span className="slider__thumb__value body-16">
            {formatRegExp ? String(value).replaceAll(formatRegExp.searchValue, formatRegExp.replaceValue) : value}
          </span>
        </div>
      </div>
      {marks && (
        <div className="slider__marks flex-row-center-center">
          {marks.map((mark, i) => {
            const rightStyle = (i === marks.length - 1 || i === 0) && zeroPercentageOnEdgeMarks;
            const { value: markValue, label } = mark;

            return (
              <span
                key={i}
                className="slider__marks__span body-14"
                style={{
                  right: `${rightStyle ? "0%" : undefined}`,
                  // It is easier to set with `position:absolute`, because we only have to set the `left` property
                  /*  
                    If we used `display: flex` and `transform: translateX(x%)` propertie, then we would have to deal
                    with each value individually. Which means, that we would need to set the `tranform` property 
                    different for the markValue < middle, the middle one and the ones with markValue < middle
                   */
                  left: `${calcPosition(mark, rightStyle)}%`,
                }}
              >
                {label !== undefined ? label : markValue}
              </span>
            );
          })}
        </div>
      )}
    </div>
  );
};

Slider.Double = Double;

export default Slider;
