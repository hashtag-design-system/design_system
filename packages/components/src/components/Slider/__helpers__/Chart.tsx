import React, { useCallback, useEffect, useRef, useState } from "react";
import { calculatePercentage } from "../../../utils";
import { useSliderContext } from "../../../utils/contexts/SliderContext";
import { useWindowDimensions } from "../../../utils/hooks";
import { SliderChartDataProp } from "../Slider";

type Props = {
  value: number | object;
};

type FProps = Props & Pick<React.ComponentPropsWithoutRef<"div">, "style">;

export const Chart: React.FunctionComponent<FProps> = ({ value, style }) => {
  const [chartOverlayLeft, setChartOverlayLeft] = useState<number>(0);
  const windowDimensions = useWindowDimensions();

  const chartOverlay = useRef<HTMLDivElement>(null);
  const chartOverlayFrameRef = useRef<HTMLDivElement>(null);
  const chartFrameRef = useRef<HTMLDivElement>(null);

  const { chart } = useSliderContext();

  const calcBarHeight = useCallback(
    (data: SliderChartDataProp) => {
      if (!chart) {
        return undefined;
      }
      const { percentage } = chart;
      const { value } = data;
      if (percentage) {
        return value;
      } else {
        const data = chart.data.map(datum => datum.value);
        const maxVal = Math.max(...data);
        const perc = calculatePercentage(value, 0, maxVal, { returnRounded: true });
        return perc;
      }
    },
    [chart]
  );

  useEffect(() => {
    if (chartOverlay && chartOverlay.current) {
      setChartOverlayLeft(chartOverlay.current.offsetLeft);
    }
  }, [chartOverlay, value]);

  useEffect(() => {
    if (chartFrameRef && chartFrameRef.current && chartOverlayFrameRef && chartOverlayFrameRef.current) {
      chartOverlayFrameRef.current.style.width = `${chartFrameRef.current.offsetWidth}px`;
    }
  }, [windowDimensions.width, chartFrameRef, chartOverlayFrameRef]);

  return chart ? (
    <div className="slider__chart flex-column-flex-start-flex-start">
      <div className="slider__chart__overlay" style={style} ref={chartOverlay}>
        <div
          className="slider__chart__frame slider__chart__frame__overlay"
          ref={chartOverlayFrameRef}
          style={{ left: `-${chartOverlayLeft}px` }}
        >
          {chart.data.map((datum, i) => {
            return <div key={i} className="slider__chart__bar" style={{ height: `${calcBarHeight(datum)}%` }}></div>;
          })}
        </div>
      </div>
      <div className="slider__chart__frame" ref={chartFrameRef}>
        {chart.data.map((datum, i) => {
          return <div key={i} className="slider__chart__bar" style={{ height: `${calcBarHeight(datum)}%` }}></div>;
        })}
      </div>
    </div>
  ) : null;
};
