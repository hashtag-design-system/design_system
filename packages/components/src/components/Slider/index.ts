import Slider from "./Slider";

export type { Props as SliderDoubleProps } from "./Double";
export type { FProps as SliderFProps, Props as SliderProps } from "./Slider";

export default Slider;
