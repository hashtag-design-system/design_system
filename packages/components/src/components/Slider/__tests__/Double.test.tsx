import { render, screen } from "@testing-library/react";
import userEvent, { specialChars } from "@testing-library/user-event";
import React from "react";
import { calculatePercentage } from "../../../utils";
import Slider from "../index";
import { TEST_CHART_DATA } from "../__helpers__/Chart/Chart.test";
import { defaultProps } from "../__helpers__/utils";

describe("<Slider.Double />", () => {
  test("default behaviour", () => {
    render(<Slider.Double />);
    const inputs = screen.getAllByTestId("slider-input");
    const container = screen.getByTestId("slider-container");
    const thumbs = screen.getAllByTestId("slider-thumb");
    const sliderField = screen.getByTestId("slider-field");

    // inputs tests
    inputs.forEach(input => {
      expect(input).toBeVisible();
      expect(input).toHaveAttribute("type", "range");
      expect(input).toHaveAttribute("class");
      expect(input).not.toBeDisabled();
      expect(input).not.toHaveAttribute("aria-disabled");
      expect(input).toHaveValue("25");
      expect(input).toHaveAttribute("min", "0");
      expect(input).toHaveAttribute("max", "100");
      expect(input).toHaveAttribute("step", "1");
      expect(input).toHaveAttribute("aria-valuemin", "0");
      expect(input).toHaveAttribute("aria-valuemax", "100");
      expect(input).toHaveAttribute("aria-valuenow", "25");
      expect(input).toHaveAttribute("tabindex", "-1");
      expect(input).toHaveAttribute("role", "slider");
    });

    // container tests
    expect(container).toBeVisible();
    expect(container.children).toHaveLength(1);
    expect(container).toHaveAttribute("class");

    // thumbs tests
    thumbs.forEach(thumb => {
      expect(thumb).toHaveAttribute("class");
      expect(thumb).not.toHaveClass("hover");

      if (thumb.className.includes("left")) {
        expect(thumb.style.left).toBe("25%");
      }
      if (thumb.className.includes("right")) {
        expect(thumb.style.right).toBe("25%");
      }
    });

    // sliderField tests
    expect(sliderField).toBeVisible();
    expect(sliderField.children).toHaveLength(5);

    expect(screen.queryByTestId("slider-marks-container")).toBeNull();
  });
  test("keyDown default functionality", async () => {
    render(<Slider.Double />);
    const inputs = screen.queryAllByTestId("slider-input");

    for (let i = 0; i < 10; i++) {
      userEvent.type(inputs[0], specialChars.arrowRight);
    }

    expect(inputs[0]).toHaveValue("35");
    expect(inputs[0]).toHaveAttribute("aria-valuenow", "35");
    expect(screen.queryAllByTestId("slider-thumb")[0].style.left).toBe("35%");

    for (let i = 0; i < 5; i++) {
      userEvent.type(inputs[0], specialChars.arrowLeft);
    }

    expect(inputs[0]).toHaveValue("30");
    expect(inputs[0]).toHaveAttribute("aria-valuenow", "30");
    expect(screen.queryAllByTestId("slider-thumb")[0].style.left).toBe("30%");

    for (let i = 0; i < 10; i++) {
      userEvent.type(inputs[1], specialChars.arrowDown);
    }

    expect(inputs[1]).toHaveValue("35");
    expect(inputs[1]).toHaveAttribute("aria-valuenow", "35");
    expect(screen.queryAllByTestId("slider-thumb")[1].style.right).toBe("35%");

    for (let i = 0; i < 5; i++) {
      userEvent.type(inputs[1], specialChars.arrowUp);
    }

    expect(inputs[1]).toHaveValue("30");
    expect(inputs[1]).toHaveAttribute("aria-valuenow", "30");
    expect(screen.queryAllByTestId("slider-thumb")[1].style.right).toBe("30%");
  });
  test('lThumb, state="focus-visible"', () => {
    render(<Slider.Double lThumb={{ state: "focus-visible" }} />);

    expect(screen.queryAllByTestId("slider-thumb")[0]).toHaveClass("focus-visible");
  });
  test('rThumb, state="focus-visible"', () => {
    render(<Slider.Double rThumb={{ state: "focus-visible" }} />);

    expect(screen.queryAllByTestId("slider-thumb")[1]).toHaveClass("focus-visible");
  });
  test('lThumb, state="hover"', async () => {
    render(<Slider.Double lThumb={{ state: "hover" }} />);
    const thumb = screen.queryAllByTestId("slider-thumb")[0];

    expect(thumb).toHaveClass("hover");
    userEvent.unhover(thumb);
    expect(thumb).toHaveClass("hover");
  });
  test('rThumb, state="hover"', async () => {
    render(<Slider.Double rThumb={{ state: "hover" }} />);
    const thumb = screen.queryAllByTestId("slider-thumb")[1];

    expect(thumb).toHaveClass("hover");
    userEvent.unhover(thumb);
    expect(thumb).toHaveClass("hover");
  });
  test('state="hover", with disabled={true}', () => {
    const { rerender } = render(<Slider.Double lThumb={{ state: "hover" }} disabled />);

    const thumbs = screen.queryAllByTestId("slider-thumb");

    thumbs.forEach(thumb => {
      expect(thumb).not.toHaveClass("hover");
    });

    rerender(<Slider.Double rThumb={{ state: "hover" }} disabled />);

    thumbs.forEach(thumb => {
      expect(thumb).not.toHaveClass("hover");
    });
  });
  test("disabled state", () => {
    const { rerender } = render(<Slider.Double lThumb={{ state: "disabled" }} />);
    const container = screen.getByTestId("slider-container");
    const inputs = screen.getAllByTestId("slider-input");

    expect(container).toHaveClass("disabled");
    inputs.forEach(input => {
      expect(input).toBeDisabled();
    });

    rerender(<Slider.Double disabled />);

    expect(container).toHaveClass("disabled");
    inputs.forEach(input => {
      expect(input).toBeDisabled();
    });

    rerender(<Slider.Double aria-disabled="true" />);

    expect(container).toHaveClass("disabled");
    inputs.forEach(input => {
      expect(input).toBeDisabled();
    });
  });
  test("with chart, overlay style", () => {
    render(<Slider.Double chart={TEST_CHART_DATA} />);
    const overlay = screen.getByTestId("slider-chart-overlay-container");

    expect(overlay).toBeVisible();
    expect(overlay.style.right).toBe(
      `${calculatePercentage(
        parseFloat(defaultProps.max!.toString()) / 4,
        parseFloat(defaultProps.min!.toString()),
        parseFloat(defaultProps.max!.toString())
      )}%`
    );
    expect(overlay.style.left).toBe(
      `${calculatePercentage(
        parseFloat(defaultProps.max!.toString()) / 4,
        parseFloat(defaultProps.min!.toString()),
        parseFloat(defaultProps.max!.toString())
      )}%`
    );
  });
  test("thumb onHover default behaviour", () => {
    render(<Slider.Double />);
    const thumbs = screen.getAllByTestId("slider-thumb");

    thumbs.forEach(thumb => {
      const initialWidth = parseFloat(thumb.style.width);
      const initialHeight = parseFloat(thumb.style.height);

      userEvent.hover(thumb);
      expect(thumb).toHaveClass("hover");
      expect(parseFloat(thumb.style.width)).toBeGreaterThan(initialWidth);
      expect(parseFloat(thumb.style.height)).toBeGreaterThan(initialHeight);

      userEvent.unhover(thumb);
      expect(thumb).not.toHaveClass("hover");
      expect(parseFloat(thumb.style.width)).toBe(initialWidth);
      expect(parseFloat(thumb.style.height)).toBe(initialHeight);
    });
  });
  test("number input related attributes", () => {
    render(<Slider.Double min={4} max={10} step={2} />);
    const inputs = screen.getAllByTestId("slider-input");

    inputs.forEach(input => {
      expect(input.getAttribute("min")).toBe("4");
      expect(input.getAttribute("aria-valuemin")).toBe("4");

      expect(input.getAttribute("max")).toBe("10");
      expect(input.getAttribute("aria-valuemax")).toBe("10");

      expect(input.getAttribute("step")).toBe("2");
    });
  });
});
