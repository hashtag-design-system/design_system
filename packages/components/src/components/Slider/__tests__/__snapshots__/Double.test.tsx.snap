// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider <Chart /> with chart={true} 1`] = `
<div
  class="slider__chart__frame"
  data-testid="slider-chart"
>
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 10%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 20%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 30%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 40%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 50%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 60%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 70%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 80%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 90%;"
  />
  <div
    class="slider__chart__bar"
    data-testid="slider-chart-bar"
    style="height: 100%;"
  />
</div>
`;
